MigrationPreparer/MigrationPreparerImplementation.cpp

```diff
@@ -0,0 +1,595 @@
+        if (list.empty()) return Core::ERROR_NONE;  // Handle empty list case
+        auto it = std::find(list.begin(), list.end(), searchString);
+        if (it != list.end()) {
+            return true;
+        } 
+        return false;
+    }
+   
+    /*Helper's: End*/
+
+    uint32_t MigrationPreparerImplementation::writeEntry(const string& name, const string &value) {
+
+        LOGINFO("[WRITE] params={name: %s, value: %s}", name.c_str(), value.c_str());
+        string entry;
+        string key = name;
+        string newValue = value;
+        WPEFramework::Core::File dataStore(DATASTORE_PATH);
+        WPEFramework::Core::Directory dataStoreDir(DATASTORE_DIR);
+        
+        // check if someone deletes the dataStore in the middle of the operation
+        if(!lineNumber.empty() && !dataStore.Exists()) {
+            LOGWARN("Migration datastore %s deleted in the middle of the operation, resetting the internal data structures", DATASTORE_PATH);
+            lineNumber.clear();
+            curLineIndex = 1;
+            if (!valueEntry.empty())
+                valueEntry.clear();
+        }
+        
+        dataStoreMutex.lock();
+        // Handle first Write request
+        if (lineNumber.empty()) {
+            curLineIndex = 1;
+            if(!dataStore.Exists()) {
+                // check if the directory exist
+                if(!dataStoreDir.Exists())
+                    // assuming /opt/secure path already exist and "migration" folder needs to be under /opt/secure/
+                    dataStoreDir.Create();
+                
+                if(!dataStore.Create()) {
+                    LOGERR("Failed to create migration datastore %s, errno: %d, reason: %s", DATASTORE_PATH, errno, strerror(errno));
+                    dataStoreMutex.unlock();
+                    return ERROR_CREATE;
+                }
+                fileExist = true;
+            }
+            
+            if(!dataStore.Open(false)) {
+                LOGERR("Failed to open migration datastore %s, errno: %d, reason: %s", DATASTORE_PATH, errno, strerror(errno));
+                LOGERR("Failed to create entry for {%s:%s} in migration datastore", key.c_str(), newValue.c_str());
+                dataStoreMutex.unlock();
+                return ERROR_OPEN;
+            }
+            // write entry to the dataStore
+            entry = string("{\n \"") + key + string("\":") + newValue + string("\n}");
+            dataStore.Write(reinterpret_cast<const uint8_t*>(&entry[0]), entry.size());
+            // update the entry in lineNumber map
+            lineNumber[key] = ++curLineIndex;
+            valueEntry.push_back(newValue);
+            dataStore.Close();
+            dataStoreMutex.unlock();
+            return Core::ERROR_NONE;
+        }
+
+        // Handle Update Request
+        if(lineNumber.find(key) != lineNumber.end()) {
+
+            string oldValue = valueEntry[lineNumber[key] - 2];
```

Comment:
A sample dataStore file would look like below
{
"key1":"value1",
"key2":"value2"
}

Line1 ---> {
Line2 ---> first item

Since "valueEntry" is a vector of strings whose index would start from 0.
We had to offset line number of items by -2 to access the values.